### PR TITLE
[ios] fix memory leak in SearchBar

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/dotnet/maui/pull/16346

This addresses the memory leak discovered by:

    src/Core/src/Platform/iOS/MauiSearchBar.cs(36,63): error MA0001: Event 'TextSetOrChanged' could cause memory leaks in an NSObject subclass. Remove the event or add the [UnconditionalSuppressMessage("Memory", "MA0001")] attribute with a justification as to why the event will not leak.
    src/Core/src/Platform/iOS/MauiSearchBar.cs(54,32): error MA0001: Event 'OnMovedToWindow' could cause memory leaks in an NSObject subclass. Remove the event or add the [UnconditionalSuppressMessage("Memory", "MA0001")] attribute with a justification as to why the event will not leak.
    src/Core/src/Platform/iOS/MauiSearchBar.cs(55,32): error MA0001: Event 'EditingChanged' could cause memory leaks in an NSObject subclass. Remove the event or add the [UnconditionalSuppressMessage("Memory", "MA0001")] attribute with a justification as to why the event will not leak.
    src/Core/src/Platform/iOS/MauiSearchBar.cs(70,31): error MA0003: Subscribing to events with instance method 'OnEditingChanged' could cause memory leaks in an NSObject subclass. Remove the subscription or convert the method to a static method.

I could reproduce a leak in a test such as:

    await InvokeOnMainThreadAsync(() =>
    {
        var layout = new Grid();
        var searchBar = new SearchBar();
        layout.Add(searchBar);
        var handler = CreateHandler<LayoutHandler>(layout);
        viewReference = new WeakReference(searchBar);
        handlerReference = new WeakReference(searchBar.Handler);
        platformViewReference = new WeakReference(searchBar.Handler.PlatformView);
    });

    await AssertionExtensions.WaitForGC(viewReference, handlerReference, platformViewReference);
    Assert.False(viewReference.IsAlive, "SearchBar should not be alive!");
    Assert.False(handlerReference.IsAlive, "Handler should not be alive!");
    Assert.False(platformViewReference.IsAlive, "PlatformView should not be alive!");

Solved the issues by making a non-NSObject `MauiSearchBarProxy` class.

I found & fixed a couple related issues:

* `SearchBarExtensions.GetSearchTextField()` would have always thrown `StackOverlowException` if iOS was less than 13? It just called itself?!?

* Removed `MauiSearchBar.EditingChanged`, as we could subscribe to the event from the `UITextField` directly instead.

Fixes - #20024

### CODE REVIEW

File: src/Core/src/Platform/iOS/SearchBarExtensions.cs [hunk 1]
```diff
@@ -10,9 +10,25 @@ public static class SearchBarExtensions
 		internal static UITextField? GetSearchTextField(this UISearchBar searchBar)
 		{
 			if (OperatingSystem.IsIOSVersionAtLeast(13))
+			{
 				return searchBar.SearchTextField;
-			else
-				return searchBar.GetSearchTextField();
+			}
```
Comment (jonathanpeppers) [inline src/Core/src/Platform/iOS/SearchBarExtensions.cs:15]: If iOS < 13 before, this just threw `StackOverflowException`? Let me know if I'm missing something, I ported a Swift example to C# instead.

File: src/Core/src/Platform/iOS/MauiSearchBar.cs [hunk 2]
```diff
@@ -51,32 +52,18 @@ protected internal MauiSearchBar(NativeHandle handle) : base(handle)
...
-			if (editor != null)
+			if (this.GetSearchTextField() is UITextField editor)
 				ResignFirstResponderTouchGestureRecognizer.Update(editor, window);
 
-			if (editor != null)
-			{
-				editor.EditingChanged -= OnEditingChanged;
-				if (window != null)
-					editor.EditingChanged += OnEditingChanged;
-			}
```
Comment (jonathanpeppers) [inline src/Core/src/Platform/iOS/MauiSearchBar.cs:71]: I removed this `EditingChanged` event and subscribe in `MauiSearchBarProxy` instead.

**Comment by jonathanpeppers** [comment]
This has UITest failures I haven't seen on other PRs, there must be a problem. Looking into it.

**Comment by jonathanpeppers** [comment]
@rmarinho I was just going to come back to this one when UITests are working.

**Comment by jonathanpeppers** [comment]
This is green, but the UITests are disabled -- I am wondering if this actually breaks something, so will wait for UITests.

**Comment by rmarinho** [comment]
@jonathanpeppers  can you rebase main?! and i can run them for you to check. Main should be green for uitest right now, we will enable during this week

File: src/Core/src/Handlers/SearchBar/SearchBarHandler.iOS.cs [hunk 3]
```diff
@@ -23,35 +24,14 @@ protected override MauiSearchBar CreatePlatformView()
...
 		{
-			platformView.CancelButtonClicked += OnCancelClicked;
-			platformView.SearchButtonClicked += OnSearchButtonClicked;
-			platformView.TextSetOrChanged += OnTextPropertySet;
-			platformView.OnMovedToWindow += OnMovedToWindow;
-			platformView.ShouldChangeTextInRange += ShouldChangeText;
-			platformView.OnEditingStarted += OnEditingStarted;
-			platformView.EditingChanged += OnEditingChanged;
-			platformView.OnEditingStopped += OnEditingStopped;
+			_proxy.Connect(this, VirtualView, platformView);
```
Comment (jsuarezruiz) [inline src/Core/src/Handlers/SearchBar/SearchBarHandler.iOS.cs:27]: LGTM.

**Review by jsuarezruiz — COMMENTED** [review]
I see that the status is Draft, is there anything pending?

File: src/Core/src/Handlers/SearchBar/SearchBarHandler.iOS.cs (hunk 3 above)
Comment (jonathanpeppers) [reply src/Core/src/Handlers/SearchBar/SearchBarHandler.iOS.cs:27]: This PR completely breaks all the UI tests, which accesses a `SearchBar` to find & run each test?

I did not figure out the cause, it seemed like `SearchBar` generally worked when I tried in other apps like the DeviceTests.

**Review by tj-devel709 — APPROVED** [review]

**Review by rmarinho — APPROVED** [review]

**Pull request merged**